src/Microsoft.ML.Data/Transforms/KeyToValueTransform.cs

```diff
@@ -67,386 +79,547 @@ private static VersionInfo GetVersionInfo()
+            {
+                Input = input;
+            }
+        }
+
+        private sealed class OutScalarColumn<TKey, TValue> : Scalar<TValue>, IColInput
+        {
+            public PipelineColumn Input { get; }
+
+            public OutScalarColumn(Key<TKey, TValue> input)
+                : base(Reconciler.Inst, input)
+            {
+                Input = input;
+            }
+        }
+
+        private sealed class OutVectorColumn<TKey, TValue> : Vector<TValue>, IColInput
+        {
+            public PipelineColumn Input { get; }
+
+            public OutVectorColumn(Vector<Key<TKey, TValue>> input)
+                : base(Reconciler.Inst, input)
+            {
+                Input = input;
+            }
+        }
+
+        private sealed class OutVarVectorColumn<TKey, TValue> : VarVector<TValue>, IColInput
+        {
+            public PipelineColumn Input { get; }
+
+            public OutVarVectorColumn(VarVector<Key<TKey, TValue>> input)
+                : base(Reconciler.Inst, input)
+            {
+                Input = input;
+            }
+        }
+
+        private sealed class Reconciler : EstimatorReconciler
+        {
+            public static Reconciler Inst = new Reconciler();
+
+            private Reconciler() { }
+
+            public override IEstimator<ITransformer> Reconcile(IHostEnvironment env,
+                PipelineColumn[] toOutput,
+                IReadOnlyDictionary<PipelineColumn, string> inputNames,
+                IReadOnlyDictionary<PipelineColumn, string> outputNames,
+                IReadOnlyCollection<string> usedNames)
+            {
+                var cols = new (string input, string output)[toOutput.Length];
+                for (int i = 0; i < toOutput.Length; ++i)
                 {
-                    var funcName = ctx.GetFreeFunctionName("mapKeyToValue");
-                    ctx.Pfa.AddFunc(funcName, new JArray(PfaUtils.Param("key", PfaUtils.Type.Int)),
-                        outType, PfaUtils.If(PfaUtils.Call("<", "key", 0), defaultToken,
-                        PfaUtils.Index(cellRef, "key")));
-                    var funcRef = PfaUtils.FuncRef("u." + funcName);
-                    return PfaUtils.Call("a.map", srcToken, funcRef);
+                    var outCol = (IColInput)toOutput[i];
+                    cols[i] = (inputNames[outCol.Input], outputNames[toOutput[i]]);
                 }
-                return PfaUtils.If(PfaUtils.Call("<", srcToken, 0), defaultToken, PfaUtils.Index(cellRef, srcToken));
+                return new KeyToValueEstimator(env, cols);
             }
         }
+
+        /// <summary>
+        /// Convert a key column to a column containing the corresponding value.
+        /// </summary>
+        public static Key<T2, T3> ToValue<T1, T2, T3>(this Key<T1, Key<T2, T3>> input)
+        {
+            Contracts.CheckValue(input, nameof(input));
+            return new OutKeyColumn<T1, T2, T3>(input);
+        }
+
```

Comment:
This is not presently possible, since you cannot have key-values that are themselves keys with key-values... you can however have key. So I think an overload for `Key<K1, Key<K2>>` would make sense, but not this.